File: src/proxy/pool.rs
use std::collections::{HashMap, HashSet, VecDeque};
use std::sync::LazyLock;
use std::sync::Mutex;
use std::time::Instant;

use crate::proxy::config::PROXY_CONFIG;
use crate::proxy::raw_websocket::RawWebSocket;
use crate::proxy::utils::{dc_default_ips, ws_domains};
use crate::proxy::STATS;

type PoolBucket<T> = VecDeque<(T, Instant)>;

pub struct WsPool {
    idle: Mutex<HashMap<(u32, bool), PoolBucket<RawWebSocket>>>,
    refilling: Mutex<HashSet<(u32, bool)>>,
}

impl Default for WsPool {
    fn default() -> Self {
        Self::new()
    }
}

impl WsPool {
    pub fn new() -> Self {
        Self {
            idle: Mutex::new(HashMap::new()),
            refilling: Mutex::new(HashSet::new()),
        }
    }

    pub async fn get(
        &self,
        dc: u32,
        is_media: bool,
        target_ip: &str,
        domains: &[String],
    ) -> Option<RawWebSocket> {
        let key = (dc, is_media);
        let now = Instant::now();

        {
            let mut idle = self.idle.lock().unwrap();
            if let Some(bucket) = idle.get_mut(&key) {
                while let Some((ws, created)) = bucket.pop_front() {
                    let age = now.duration_since(created).as_secs_f64();
                    if age > 120.0 || ws.is_closed() {
                        tokio::spawn(quiet_close(ws));
                        continue;
                    }
                    STATS
                        .pool_hits
                        .fetch_add(1, std::sync::atomic::Ordering::Relaxed);
                    drop(idle);
                    self.schedule_refill(key, target_ip, domains);
                    return Some(ws);
                }
            }
        }

        STATS
            .pool_misses
            .fetch_add(1, std::sync::atomic::Ordering::Relaxed);
        self.schedule_refill(key, target_ip, domains);
        None
    }

    fn schedule_refill(&self, key: (u32, bool), target_ip: &str, domains: &[String]) {
        let mut refilling = self.refilling.lock().unwrap();
        if refilling.contains(&key) {
            return;
        }
        refilling.insert(key);

        let target_ip = target_ip.to_string();
        let domains = domains.to_vec();

        tokio::spawn(async move {
            let pool = &WS_POOL;
            let pool_size = {
                let cfg = PROXY_CONFIG.read().unwrap();
                cfg.pool_size
            };

            let needed = {
                let mut idle = pool.idle.lock().unwrap();
                let bucket = idle.entry(key).or_default();
                pool_size.saturating_sub(bucket.len())
            };

            if needed > 0 {
                let mut handles = Vec::new();
                for _ in 0..needed {
                    let tip = target_ip.clone();
                    let doms = domains.clone();
                    handles.push(tokio::spawn(async move {
                        connect_one(&tip, &doms).await
                    }));
                }
                for handle in handles {
                    if let Ok(Some(ws)) = handle.await {
                        let mut idle = pool.idle.lock().unwrap();
                        idle.entry(key).or_default().push_back((ws, Instant::now()));
                    }
                }
            }

            pool.refilling.lock().unwrap().remove(&key);
        });
    }

    pub async fn warmup(&self) {
        let dc_redirects = {
            let cfg = PROXY_CONFIG.read().unwrap();
            cfg.dc_redirects.clone()
        };
        for (dc, target_ip) in &dc_redirects {
            for is_media in [false, true] {
                let domains = ws_domains(*dc, is_media);
                self.schedule_refill((*dc, is_media), target_ip, &domains);
            }
        }
    }

    pub fn reset(&self) {
        self.idle.lock().unwrap().clear();
        self.refilling.lock().unwrap().clear();
    }
}

pub struct CfWorkerPool {
    idle: Mutex<HashMap<(u32, String), PoolBucket<RawWebSocket>>>,
    refilling: Mutex<HashSet<(u32, String)>>,
}

impl Default for CfWorkerPool {
    fn default() -> Self {
        Self::new()
    }
}

impl CfWorkerPool {
    pub fn new() -> Self {
        Self {
            idle: Mutex::new(HashMap::new()),
            refilling: Mutex::new(HashSet::new()),
        }
    }

    pub async fn get(
        &self,
        dc: u32,
        worker_domain: &str,
        fallback_dst: &str,
    ) -> Option<RawWebSocket> {
        let key = (dc, worker_domain.to_string());
        let now = Instant::now();

        {
            let mut idle = self.idle.lock().unwrap();
            if let Some(bucket) = idle.get_mut(&key) {
                while let Some((ws, created)) = bucket.pop_front() {
                    let age = now.duration_since(created).as_secs_f64();
                    if age > 120.0 || ws.is_closed() {
                        tokio::spawn(quiet_close(ws));
                        continue;
                    }
                    STATS
                        .cf_pool_hits
                        .fetch_add(1, std::sync::atomic::Ordering::Relaxed);
                    drop(idle);
                    self.schedule_refill(&key, fallback_dst);
                    return Some(ws);
                }
            }
        }

        STATS
            .cf_pool_misses
            .fetch_add(1, std::sync::atomic::Ordering::Relaxed);
        self.schedule_refill(&key, fallback_dst);
        None
    }

    fn schedule_refill(&self, key: &(u32, String), fallback_dst: &str) {
        let mut refilling = self.refilling.lock().unwrap();
        if refilling.contains(key) {
            return;
        }
        refilling.insert(key.clone());

        let key = key.clone();
        let fallback_dst = fallback_dst.to_string();

        tokio::spawn(async move {
            let pool = &CF_WORKER_POOL;
            let pool_size = {
                let cfg = PROXY_CONFIG.read().unwrap();
                cfg.pool_size
            };

            let needed = {
                let mut idle = pool.idle.lock().unwrap();
                let bucket = idle.entry(key.clone()).or_default();
                pool_size.saturating_sub(bucket.len())
            };

            if needed > 0 {
                let (dc, worker_domain) = &key;
                let path = format!("/apiws?dst={}&dc={}", fallback_dst, dc);

                let mut handles = Vec::new();
                for _ in 0..needed {
                    let wd = worker_domain.clone();
                    let p = path.clone();
                    handles.push(tokio::spawn(async move {
                        RawWebSocket::connect(&wd, &wd, std::time::Duration::from_secs(8), &p)
                            .await
                            .ok()
                    }));
                }

                for handle in handles {
                    if let Ok(Some(ws)) = handle.await {
                        let mut idle = pool.idle.lock().unwrap();
                        idle.entry(key.clone())
                            .or_default()
                            .push_back((ws, Instant::now()));
                    }
                }
            }

            pool.refilling.lock().unwrap().remove(&key);
        });
    }

    pub async fn warmup(&self) {
        let cf_worker_domains = {
            let cfg = PROXY_CONFIG.read().unwrap();
            cfg.cfproxy_worker_domains.clone()
        };
        if cf_worker_domains.is_empty() {
            return;
        }
        let dc_redirects = {
            let cfg = PROXY_CONFIG.read().unwrap();
            cfg.dc_redirects.clone()
        };
        let cf_fallbacks: HashMap<u32, String> = dc_default_ips()
            .into_iter()
            .filter(|(dc, _)| !dc_redirects.contains_key(dc))
            .map(|(k, v)| (k, v.to_string()))
            .collect();

        for worker_domain in &cf_worker_domains {
            for (dc, fallback_dst) in &cf_fallbacks {
                self.schedule_refill(&(*dc, worker_domain.clone()), fallback_dst);
            }
        }
    }

    pub fn reset(&self) {
        self.idle.lock().unwrap().clear();
        self.refilling.lock().unwrap().clear();
    }
}

async fn connect_one(target_ip: &str, domains: &[String]) -> Option<RawWebSocket> {
    for domain in domains {
        match RawWebSocket::connect(
            target_ip,
            domain,
            std::time::Duration::from_secs(8),
            "/apiws",
        )
        .await
        {
            Ok(ws) => return Some(ws),
            Err(e) => {
                if e.downcast_ref::<crate::proxy::raw_websocket::WsHandshakeError>()
                    .is_some_and(|ws_err| ws_err.is_redirect())
                {
                    continue;
                }
                return None;
            }
        }
    }
    None
}

async fn quiet_close(mut ws: RawWebSocket) {
    ws.close().await;
}

pub static WS_POOL: LazyLock<WsPool> = LazyLock::new(WsPool::new);
pub static CF_WORKER_POOL: LazyLock<CfWorkerPool> = LazyLock::new(CfWorkerPool::new);

File: proxy/pool.py
import asyncio
import logging
import time

from collections import deque
from urllib.parse import urlencode
from typing import Any, Coroutine, Dict, List, Optional, Tuple, Set

from .raw_websocket import RawWebSocket, WsHandshakeError
from .stats import stats
from .config import proxy_config
from .utils import ws_domains, DC_DEFAULT_IPS

log = logging.getLogger('tg-mtproto-proxy')

WS_POOL_MAX_AGE = 120.0


class _BasePool:
    """Generic connection pool keyed by an arbitrary hashable key.

    Subclasses only need to implement:
      - ``_connect_one(...)`` — create a single connection.
      - ``_refill_args(key)`` — return keyword arguments for ``_connect_one``.
      - ``_log_prefix(key)`` — human-readable label for log lines.
      - ``_inc_hit()`` / ``_inc_miss()`` — bump the right stats counter.
    """

    def __init__(self, key_type=Any):
        self._idle: Dict[Any, deque] = {}
        self._refilling: Set[Any] = set()

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    async def get(self, key, *args, **kwargs) -> Optional[RawWebSocket]:
        now = time.monotonic()
        bucket = self._idle.get(key)
        if bucket is None:
            bucket = deque()
            self._idle[key] = bucket

        while bucket:
            ws, created = bucket.popleft()
            age = now - created
            if age > WS_POOL_MAX_AGE or ws._closed or ws.writer.transport.is_closing():
                asyncio.create_task(self._quiet_close(ws))
                continue
            self._inc_hit()
            log.debug("%s pool hit %s (age=%.1fs, left=%d)",
                      self._log_prefix(key), key, age, len(bucket))
            self._schedule_refill(key, *args, **kwargs)
            return ws

        self._inc_miss()
        self._schedule_refill(key, *args, **kwargs)
        return None

    async def warmup(self) -> None:
        tasks = self._schedule_all_refills()
        if tasks:
            await asyncio.wait(tasks)
            log.info("%s pool warmup complete (%d DC(s))",
                     self._name, self._count_keys())
        else:
            log.info("%s pool warmup: nothing to do", self._name)

    def reset(self) -> None:
        self._idle.clear()
        self._refilling.clear()

    # ------------------------------------------------------------------
    # Internals — subclasses override the hooks
    # ------------------------------------------------------------------

    _name: str = "base"

    def _schedule_all_refills(self) -> List[asyncio.Task]:
        raise NotImplementedError

    def _connect_one(self, *args, **kwargs) -> Coroutine:
        raise NotImplementedError

    def _refill_args(self, key) -> Tuple[tuple, dict]:
        raise NotImplementedError

    def _log_prefix(self, key) -> str:
        return self._name

    def _count_keys(self) -> int:
        raise NotImplementedError

    def _inc_hit(self) -> None:
        pass

    def _inc_miss(self) -> None:
        pass

    # ------------------------------------------------------------------
    # Shared mechanics
    # ------------------------------------------------------------------

    def _schedule_refill(self, key, *args, **kwargs) -> None:
        if key in self._refilling:
            return
        self._refilling.add(key)
        asyncio.create_task(self._refill(key, *args, **kwargs))

    async def _refill(self, key, *args, **kwargs) -> None:
        try:
            bucket = self._idle.setdefault(key, deque())
            needed = proxy_config.pool_size - len(bucket)
            if needed <= 0:
                return
            args, kwargs = self._refill_args(key)
            connect_tasks = [
                asyncio.create_task(self._connect_one(*args, **kwargs))
                for _ in range(needed)
            ]
            for t in connect_tasks:
                try:
                    ws = await t
                    if ws:
                        bucket.append((ws, time.monotonic()))
                except Exception:
                    pass
            log.debug("%s pool refilled %s: %d ready",
                      self._log_prefix(key), key, len(bucket))
        finally:
            self._refilling.discard(key)

    @staticmethod
    async def _quiet_close(ws: RawWebSocket) -> None:
        try:
            await ws.close()
        except Exception:
            pass


class _WsPool(_BasePool):
    _name = "WS"

    def __init__(self):
        super().__init__()

    def _inc_hit(self) -> None:
        stats.pool_hits += 1

    def _inc_miss(self) -> None:
        stats.pool_misses += 1

    def _log_prefix(self, key) -> str:
        dc, is_media = key
        return f"DC{dc}{'m' if is_media else ''}"

    def _count_keys(self) -> int:
        return len(proxy_config.dc_redirects)

    def _refill_args(self, key) -> Tuple[tuple, dict]:
        dc, is_media = key
        target_ip = proxy_config.dc_redirects[dc]
        domains = ws_domains(dc, is_media)
        return (target_ip, domains), {}

    async def _connect_one(self, target_ip: str, domains: List[str]) -> Optional[RawWebSocket]:
        for domain in domains:
            try:
                return await RawWebSocket.connect(target_ip, domain, timeout=8)
            except WsHandshakeError as exc:
                if exc.is_redirect:
                    continue
                return None
            except Exception:
                return None
        return None

    def _schedule_all_refills(self) -> List[asyncio.Task]:
        tasks = []
        for dc, target_ip in proxy_config.dc_redirects.items():
            if target_ip is None:
                continue
            for is_media in (False, True):
                key = (dc, is_media)
                if key in self._refilling:
                    continue
                self._refilling.add(key)
                tasks.append(asyncio.create_task(self._refill(key, target_ip, ws_domains(dc, is_media))))
        log.info("WS pool warmup started for %d DC(s)", self._count_keys())
        return tasks


class _CfWorkerPool(_BasePool):
    _name = "CF"

    def __init__(self):
        super().__init__()

    def _inc_hit(self) -> None:
        stats.cf_pool_hits += 1

    def _inc_miss(self) -> None:
        stats.cf_pool_misses += 1

    def _log_prefix(self, key) -> str:
        dc, worker_domain = key
        return f"DC{dc}"

    def _count_keys(self) -> int:
        cf_fallbacks = {
            dc for dc in DC_DEFAULT_IPS
            if dc not in proxy_config.dc_redirects
        }
        return len(cf_fallbacks)

    def _refill_args(self, key) -> Tuple[tuple, dict]:
        dc, worker_domain = key
        fallback_dst = DC_DEFAULT_IPS.get(dc, '')
        return (worker_domain, fallback_dst, dc), {}

    async def _connect_one(self, worker_domain: str, fallback_dst: str, dc: int) -> Optional[RawWebSocket]:
        query = urlencode({'dst': fallback_dst, 'dc': str(dc)})
        path = f'/apiws?{query}'
        try:
            return await RawWebSocket.connect(worker_domain, worker_domain, timeout=8, path=path)
        except Exception:
            return None

    def _schedule_all_refills(self) -> List[asyncio.Task]:
        cf_fallbacks = {
            dc: ip for dc, ip in DC_DEFAULT_IPS.items()
            if dc not in proxy_config.dc_redirects
        }
        if not cf_fallbacks or not proxy_config.cfproxy_worker_domains:
            return []

        tasks = []
        for worker_domain in proxy_config.cfproxy_worker_domains:
            for dc, fallback_dst in cf_fallbacks.items():
                key = (dc, worker_domain)
                if key in self._refilling:
                    continue
                self._refilling.add(key)
                tasks.append(asyncio.create_task(self._refill(key, worker_domain, fallback_dst, dc)))
        log.info("CF worker pool warmup started for %d DC(s)", len(cf_fallbacks))
        return tasks


ws_pool = _WsPool()
cf_worker_pool = _CfWorkerPool()
